cpu/mips32r2_common/thread_arch.c

```diff
@@ -0,0 +1,399 @@
+/*
+ * Copyright 2016, Imagination Technologies Limited and/or its
+ *                 affiliated group companies.
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+#include <mips/cpu.h>
+#include <mips/hal.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "thread.h"
+#include "cpu.h"
+#include "irq.h"
+#include "cpu_conf.h"
+#include "periph_conf.h" /* for debug uart number */
+#include "periph/uart.h"
+#include "malloc.h"
+
+#define STACK_END_PAINT    (0xdeadc0de)
+#define C0_STATUS_EXL      (2)
+#define PADDING            (16)
+#define MICROMIPS_ISA_MODE (1)
+#define M32_SYSCALL        (0xC)
+#define M32_SYSCALL_MASK   (0xfc00003f)
+
+/* 
+ * note the major 16bits of a 32bit MicroMIPS opcode appear first in the
+ * instruction stream
+ */
+#define MM_SYSCALL         (0x8B7C0000)
+#define MM_SYSCALL_MASK    (0xfffffc00)
+
+
+#ifdef MIPS_HARD_FLOAT
+/* pointer to the current and old fpu context for lazy context switching */
+static struct fp64ctx *currentfpctx;   /* fpu context of current running task */
+static struct fp64ctx *oldfpctx;       /* fpu context of last task that executed fpu */
+#endif
+
+/*
+ *    Stack Layout, note struct gpctx is defined in
+ *    $MIPS_ELF_ROOT/mips-mti-elf/include/mips/hal.h
+ *
+ *    Top Of Stack
+ *     ---------------
+ *    |               |
+ *    |  User stack   |
+ *    |               |
+ *     ---------------  <--- gpctx->sp
+ *    |               |
+ *    |    gpctx      |
+ *    |               |
+ *     ---------------
+ *    |  16 byte pad  |
+ *     ---------------   <--- sched_active_thread->sp
+ */
+
+char *thread_arch_stack_init(thread_task_func_t task_func, void *arg,
+                             void *stack_start, int stack_size)
+{
+    /* make sure it is aligned to 8 bytes this is a requirement of the O32 ABI */
+    uintptr_t *p = (uintptr_t *)((long)(stack_start + stack_size) & ~7);
+    uintptr_t *fp;
+
+    /* paint */
+    p--;
+    *p-- = STACK_END_PAINT;
+
+
```

Comment:
?? where ? Do you want the double newline removing #74 or one adding between #71 and #72 ?